tests/unit/test_charm.py

```diff
@@ -1,18 +1,1264 @@
+            networks={"juju-info": Network.default()},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert out.unit_status == ops.WaitingStatus("container not yet connectable")
+
+    def test_given_only_database_storage_container_cant_connect_network_available_gocert_not_running_when_collect_status_then_status_is_waiting(
+        self, context
+    ):
+        state = State(
+            storage=[Storage(name="database")],
+            containers=[Container(name="gocert", can_connect=False)],
+            networks={"juju-info": Network.default()},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert out.unit_status == ops.WaitingStatus("container not yet connectable")
+
+    def test_given_storages_available_container_cant_connect_network_available_gocert_not_running_when_collect_status_then_status_is_waiting(
+        self, context
+    ):
+        state = State(
+            storage=[Storage(name="config"), Storage(name="database")],
+            containers=[Container(name="gocert", can_connect=False)],
+            networks={"juju-info": Network.default()},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert out.unit_status == ops.WaitingStatus("container not yet connectable")
+
+    def test_given_only_config_storage_container_can_connect_network_available_gocert_not_running_when_collect_status_then_status_is_waiting(
+        self, context
+    ):
+        state = State(
+            storage=[Storage(name="config")],
+            containers=[Container(name="gocert", can_connect=True)],
+            networks={"juju-info": Network.default()},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert out.unit_status == ops.WaitingStatus("storages not yet available")
+
+    def test_given_only_database_storage_container_can_connect_network_available_gocert_not_running_when_collect_status_then_status_is_waiting(
+        self, context
+    ):
+        state = State(
+            storage=[Storage(name="database")],
+            containers=[Container(name="gocert", can_connect=True)],
+            networks={"juju-info": Network.default()},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert out.unit_status == ops.WaitingStatus("storages not yet available")
```

Comment:
I think we should parametrize tests that have the same assertions. For example I don't think the test for the db storage is available vs the config one can be combined into 1 (with parameters) as the expected behavior from our code is the same for both.

This may apply elsewhere, we probably don't want 3 different tests for `container not yet connectable`

Reply:
>I think we should parametrize tests that have the same assertions

I've tried doing this and in a short summary:
The issue with it is we have parameter combinations that will produce different results. This in turn results in us having to generalize the parameter to the entire state of the charm (a single parameter that includes the state of storage, containers, relations, workload etc.).

## There are 2 viable options to do parametrization:
1) The parameters are over the entire state of the charm like suggested, and include the expected assertion (it could be a status, a file name, or just straight up code)

The pros are:
* The tests look "cleaner" and easier to read since there is no branching
* The written tests to generated tests ratio is higher. This means there is less tests generated, which requires less reasoning about the parameters

The cons are:
* Requires the user to enumerate the entire state of the charm one by one. 

This is extremely similar to writing a new test for each parameter. We lose the very powerful parameter matrix ability of pytest that generates all possible combinations of the world state

* Since the world state is hand written, there will be gaps in the possibilities.
An outcome of the above con is the fact that we may skip over some world states when typing the parameters themselves that may turn out to be a problem (like storage not attached and gocert available. This was the cause of [this issue](https://github.com/canonical/vault-k8s-operator/issues/242) in Vault).

* It is very difficult to enumerate the state of the charm
Due to the nature of parameters in pytests, they all need to share the exact structure. We may want to enumerate a state that has storages, container and network binding, but if we do we must now include this in all enumerated parameters for the test. We will no longer have the option to just add a relation parameter, we must go back and add that field to every single param we described. This severely limits the benefits of having parameters, since copying and pasting the test is now more flexible than the parameter option.

2) The paremeters are over the possible states of each dependency like in pr #10 , and include their status in enumerated form
The pros are:
* We no longer have to think about the entire state of the world when developing the charm. 
When typing parameters, all we have to write is the dependency we want and its possible states. 
* The tests are more extensible
For example, all we need to do to test all of our handlers when we have a new relation added is to add a new parameter with 3 options: "Relation doesnt exist, only 1 exists, and multiple exist", and pytest will automatically generate all the new possible input states. Even if we don't add any assertions, the tests failing due to a raised error is immediately useful for charms since we do not want to raise anything in a charm.
* The files are much smaller
Even if the readability takes a hit, reading ~100 lines of parametrized tests end up being faster than reading ~1000 lines of code with many duplicates. I personally end up skimming and miss crucial minor differences when writing tests.
* the tests are fully exhaustive.
With this method, tests are exhaustive by default, and doesn't require the user to think about the states of the world at all. Just write assertions for the happy path, and trust that no error will be raised when the charm is deployed with any combination of relations, storages, connections and stages.

The cons are:
* It is a different way of reading and writing tests, which requires a change in thinking and reading
We are used to reading the title format that we have, and easily validate that the test does what the title says. If the title doesn't contain some dependency like a `peer-relation`, it is common that we do not check if that test initializes this relation because we cant know just by looking at it that it would require it. In the proposed way, we would not have to pay attention to these dependencies, and trust that as long as the parameter is exhaustive, we will test it in every possible way.

When reading and trying to understand what a test does, we now have to focus on the possible output states are instead, which aligns well with how we think about writing charms. 

* The test itself will have a bunch of branches which make it harder to read.
When evaluating this con, I frequently found myself thinking about the ratio of expected output states and the possible input states. The smaller the ratio, the better the proposed method becomes. In the case of gocert at this point, this ratio is about 7/72: 7 output states to 72 different input states, and the input states grow exponentially when the output states grow linearly. This function also bodes well to how well the approach scales as the charm grows in functionality. When a single relation and the connection to the model is added, this ratio shrinks to ~15/432. Since we also have no problem reading long branching functions in places like the `_configure` functions, I feel that we as a team are in a better position to read these types of tests.

There is actually a way to combine the two ways with pytest's metaprogramming features, but frankly that is too much for now.

all of this and examples will be a part of the spec

Reply:
Thank you for putting the effort into this, let's discuss this in the spec